network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/extensions/VoltageRegulationImpl.java

```diff
@@ -0,0 +1,113 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(VoltageRegulationImpl.class);
+
+    public VoltageRegulationImpl(Battery battery) {
+        super(battery);
+        TerminalImpl<?> terminal = (TerminalImpl<?>) getRegulatingTerminal();
+        if (terminal != null) {
+            terminal.getReferrerManager().register(this);
+        }
+    }
+
+    private BatteryImpl getBattery() {
+        return (BatteryImpl) getExtendable();
+    }
+
+    private VoltageRegulationAttributes getVoltageRegulationAttributes() {
+        return (VoltageRegulationAttributes) getBattery().getResource().getAttributes().getExtensionAttributes().get(VoltageRegulation.NAME);
+    }
+
+    @Override
+    public boolean isVoltageRegulatorOn() {
+        return getVoltageRegulationAttributes().isVoltageRegulatorOn();
+    }
+
+    @Override
+    public void setVoltageRegulatorOn(boolean voltageRegulatorOn) {
+        boolean oldValue = isVoltageRegulatorOn();
+        if (oldValue != voltageRegulatorOn) {
+            getBattery().updateResourceExtension(this,
+                res -> getVoltageRegulationAttributes().setVoltageRegulatorOn(voltageRegulatorOn),
+                "voltageRegulatorOn", oldValue, voltageRegulatorOn);
+        }
+    }
+
+    @Override
+    public double getTargetV() {
+        return getVoltageRegulationAttributes().getTargetV();
+    }
+
+    @Override
+    public void setTargetV(double targetV) {
+        double oldValue = getTargetV();
+        if (oldValue != targetV) {
+            getBattery().updateResourceExtension(this,
+                res -> getVoltageRegulationAttributes().setTargetV(targetV),
+                "targetV", oldValue, targetV);
+        }
+    }
+
+    @Override
+    public Terminal getRegulatingTerminal() {
+        return TerminalRefUtils.getTerminal(getBattery().getIndex(), getVoltageRegulationAttributes().getRegulatingTerminal());
+    }
+
+    @Override
+    public void setRegulatingTerminal(Terminal regulatingTerminal) {
+        Terminal oldValue = getRegulatingTerminal();
+        Terminal newRegulatingTerminal = regulatingTerminal != null ? regulatingTerminal : getExtendable().getTerminal();
+        if (oldValue != newRegulatingTerminal) {
+            TerminalRefAttributes finalTerminal = TerminalRefUtils.getTerminalRefAttributes(newRegulatingTerminal);
+            getBattery().updateResourceExtension(this,
+                res -> getVoltageRegulationAttributes().setRegulatingTerminal(finalTerminal),
+                "regulatingTerminal", oldValue, finalTerminal);
+        }
+
+    }
+
+    @Override
+    public void onReferencedRemoval(Terminal removedTerminal) {
+        LOGGER.warn("Change 'VoltageRegulation' regulatingTerminal to local for battery '{}', because its regulating terminal has been removed", getExtendable().getId());
+        if (removedTerminal != null && removedTerminal.equals(getRegulatingTerminal())) {
```

Comment:
`if (removedTerminal != null && removedTerminal.getConnectable().getId().equals(getRegulatingTerminal().getConnectable().getId()))`

I think it's safer to test equality with the connectable id